src/DataStructures/Tensor/Expressions/SquareRoot.hpp

```diff
@@ -18,23 +19,27 @@ namespace TensorExpressions {
 /// \brief Defines the tensor expression representing the square root of a
 /// tensor expression that evaluates to a rank 0 tensor
 ///
+/// \details The expression can have a non-zero number of indices as long as
+/// all indices are concrete time indices, as this represents a rank 0 tensor.
+///
 /// \tparam T the type of the tensor expression of which to take the square
 /// root
-template <typename T>
+/// \tparam Args the TensorIndexs of the expression
+template <typename T, typename... Args>
 struct SquareRoot
-    : public TensorExpression<SquareRoot<T>, typename T::type, tmpl::list<>,
-                              tmpl::list<>, tmpl::list<>> {
+    : public TensorExpression<SquareRoot<T, Args...>, typename T::type,
+                              typename T::symmetry, typename T::index_list,
+                              tmpl::list<Args...>> {
   static_assert(
-      std::is_base_of<TensorExpression<T, typename T::type, tmpl::list<>,
-                                       tmpl::list<>, tmpl::list<>>,
-                      T>::value,
+      (... and tt::is_time_index<Args>::value),
       "Can only take the square root of a tensor expression that evaluates to "
       "a rank 0 tensor.");
+
   using type = typename T::type;
-  using symmetry = tmpl::list<>;
```

Comment:
I would argue the symmetry, etc. should stay at `list<>`. Why would that change? I'm a bit confused by that. I would not expect the time indices to count towards rank, etc. You might need them for internal bookkeeping, but I don't think tehy should be exposed

Reply:
I thought about this for a looooong time - whether or not to propagate time indices up the tree - and eventually decided to do so for all TE types (some already merged with this decision, including `OuterProduct`, `TensorContract`, and `TensorAsExpression`). Below are my reasons for doing this, but *please* let me know if either it's not clear or if you think I'm incorrect/misguided about something or have overlooked something! In short, I think having all TE types propagate the time indices up should make the TEs  faster and less complex (?)

(1) Multi-indices (arrays) *without* time indices that are sent down the tree eventually to `TensorAsExpresion` leaves representing `Tensor`s *with* time indices means that in `TensorAsExpression::get`, we will need to make new multi-index arrays with time index slots added and copy over the input multi-index's spatial components. My thought was that the higher up in the tree we can do this, the faster and less complex the code will be, and we can push this work up higher in the tree by propagating time indices in TEs up to parent TEs (exception is that `AddSub` just propagates up index information of the first operand, not both). An example case to compare keeping vs. removing time indices is `T(ti_t, ti_a, ti_b) + (R(ti_a, ti_t, ti_b, ti_t) + S(ti_b, ti_t, ti_t, ti_a))`, where this array creation and copying of spatial index values would happen higher / less often(?) if the root `AddSub` included time indices instead of removing them and doing this array creation + copying work extra times in the `TensorAsExpression::get` leaves. With respect to the `SquareRoot` TE specifically, this propagates up time indices just to be consistent with all other TE types to allow a continued flow of time indices up the tree with nested expressions involving `sqrt`. The idea is to have all TEs do this so that we can push this work up the tree as high as possible.

(2) We have slightly more `tmpl` stuff to do if we have to remove the time indices as we build the tree up, e.g. we have to `tmpl::erase` time indices from `symmetry`, `index_list`, and `args_list`, and if we simply `tmpl::erase` time indices from `symmetry`, we have to then re-canonicalize the result since we may be left with a non-canonical symmetry after removing time indices, e.g. `R(ti_a, ti_b, ti_t)` symmetry going from `{2, 2, 1}` -> `{2, 2}` with removal, but should then be canonicalized to `{1, 1}`.

Reply:
Okay, sounds reasonable to me. Thanks!